lib/auth/auth_with_roles.go

```diff
@@ -4761,6 +4764,9 @@ func (a *ServerWithRoles) KeepAliveSemaphoreLease(ctx context.Context, lease typ
 	if err := a.action(apidefaults.Namespace, types.KindSemaphore, types.VerbUpdate); err != nil {
 		return trace.Wrap(err)
 	}
+	if a.hasBuiltinRole(types.RoleDiscovery) && lease.SemaphoreKind != types.KindAccessGraph {
+		return trace.AccessDenied("discovery service can not create other semaphores")
```

Comment:
It's not clear "other than what" semaphores, we should say other than what type or just say that this kind is forbidden.

```suggestion
		return trace.AccessDenied("discovery service can not create semaphore of kind %q", params.SemaphoreKind)
```